src/devices/src/virtio/block/device.rs

```diff
@@ -94,6 +92,8 @@ pub struct Block {
     pub(crate) queue_evt: EventFd,
     mem: GuestMemory,
 
+    device_activated: bool,
```

Comment:
I would look at this from another perspective.
A device driver normally does its all internal initialization work and when is ready to receive data does the final "registration".
Here is case where the device is receiving data while is not ready and maybe this can be fixed by how the operations are sequenced.

One idea would be to create the TAP network device but not register it or mark it as IFF_DOWN and register it or make it IFF_UP when all is ready.

Reply:
I hear you @raduiliescu, but I think that'd just make things more complicated than necessary. One of the simplifying assumptions that allowed us to simplify our architecture was that we register all events from the emulation thread.

> One idea would be to create the TAP network device but not register it or mark it as IFF_DOWN and register it or make it IFF_UP when all is ready.

- Registering the TAP fd with the event handler (listening for incoming packets) **after** the device is fully activated by the guest driver would mean doing said registration on a vcpu thread (vcpu threads do the virtio configuration since it's driven by the guest driver). We could do this by either:
  - wrap EventManager in `Arc<Mutex>` and keep references on both `vmm` and `vcpu` threads.
  - dispatch new "DeviceRegistrationComplete" events from the `vcpu` to the `vmm`, which will lead `vmm` to do the plugging of the device in EventManager.
- I believe marking the interface `IFF_DOWN/UP` might not be possible from within the jailed firecracker, even if it would, it's weird user experience to add a TAP to a microVM and have it flap up and down for a while out of your control.

Reply:
I agree with @acatangiu. We are definitely moving away from that pattern to simplify. Previously we were using a channel to inject the fds and handlers into the VMM thread.

I do not think the user would get a weird experience. It would make the experience more consistent if status of the tap device match the status of the guest nic.

Reply:
I was referring to the "register" that host TAP netdev is doing to the kernel when you create that interface, not to the register that Firecracker virtio net device is doing internally to the epoll handler. Anyway register was a sort of bad word :).
My idea here is to create the TAP, but don't let it receive packets. Allow it to receive packets when our driver is ready. One of the ways to do this is to set !IFF_UP when creating the TAP and set IFF_UP when Firecracker activates its virtio net device.